File: specs/ralph-loop/references/internal/ralphloop/prompts.go
package ralphloop

import (
	"fmt"
	"path/filepath"
	"strings"
)

type setupPromptOptions struct {
	UserPrompt   string
	PlanPath     string
	WorktreePath string
	WorktreeID   string
	WorkBranch   string
	BaseBranch   string
}

type codingPromptOptions struct {
	UserPrompt string
	PlanPath   string
}

type prPromptOptions struct {
	PlanPath   string
	BaseBranch string
}

func defaultPlanFilename(prompt string) string {
	slug := slugifyPrompt(planFilenameSeed(prompt))
	if len(slug) > 80 {
		slug = slug[:80]
		slug = strings.Trim(slug, "-")
	}
	if slug == "" {
		slug = "task"
	}
	return slug + ".md"
}

func planFilenameSeed(prompt string) string {
	issueIdentifier := ""
	issueTitle := ""
	for _, line := range strings.Split(prompt, "\n") {
		trimmed := strings.TrimSpace(line)
		switch {
		case strings.HasPrefix(trimmed, "Issue:"):
			issueIdentifier = strings.TrimSpace(strings.TrimPrefix(trimmed, "Issue:"))
		case strings.HasPrefix(trimmed, "Title:"):
			issueTitle = strings.TrimSpace(strings.TrimPrefix(trimmed, "Title:"))
		}
	}
	switch {
	case issueIdentifier != "" && issueTitle != "":
		return issueIdentifier + " " + issueTitle
	case issueIdentifier != "":
		return issueIdentifier
	default:
		return prompt
	}
}

func buildSetupPrompt(options setupPromptOptions) string {
	planName := strings.TrimSuffix(filepath.Base(options.PlanPath), filepath.Ext(options.PlanPath))
	return fmt.Sprintf(`You are the setup agent for an automated coding loop. The worktree environment is already initialized.

Task: %s

Prepared environment:
- Worktree path: %s
- Worktree ID: %s
- Working branch: %s
- Base branch: %s

Do the following steps in order:

1. Read AGENTS.md, ARCHITECTURE.md, docs/PLANS.md, and any other relevant repository docs needed to understand the task.
2. Create an execution plan at %s using the checked-in plan structure:
   - Goal / scope
   - Background
   - Milestones
   - Current progress
   - Key decisions
   - Remaining issues / open questions
   - Links to related documents
3. Break the work into 3-7 concrete milestones, each small enough for one coding-loop iteration.
4. Mark every milestone as not started.
5. Stage and commit the new plan with message: plan: %s
6. Print the absolute plan file path.

Output <promise>COMPLETE</promise> when done.`, options.UserPrompt, options.WorktreePath, options.WorktreeID, options.WorkBranch, options.BaseBranch, options.PlanPath, planName)
}

func buildCodingPrompt(options codingPromptOptions) string {
	return fmt.Sprintf(`You are a coding agent working in an automated loop. You will iterate until the task is fully complete.

## Task
%s

## Execution plan
Read the plan at %s to understand the milestones and current progress. Pick up where the last iteration left off.

## Rules
- One milestone per iteration. Complete exactly one milestone, then stop.
- Work through milestones sequentially.
- After completing the milestone, update the plan file with progress, decisions, and remaining issues.
- Stage and commit all changes from the iteration, including the plan update.
- If blocked, document the blocker in the plan and commit the current state.

## Completion signal
When all milestones are done:
- Perform the final plan update.
- Commit all remaining changes.
- Output <promise>COMPLETE</promise>.

If work remains after this iteration, do not output the completion token.`, options.UserPrompt, options.PlanPath)
}

func buildRecoveryPrompt(planPath string) string {
	return fmt.Sprintf("The previous iteration failed. Review the current git state, inspect the last errors, repair the workspace, and continue from the plan at %s. Commit the recovery work before stopping.", planPath)
}

func buildPrPrompt(options prPromptOptions) string {
	return fmt.Sprintf(`You are a PR agent. Create a pull request for the completed work on this branch.

Instructions:
1. Read the completed plan at %s.
2. Review commits with git log %s..HEAD --oneline.
3. Review scope with git diff %s...HEAD --stat.
4. Move the plan from docs/exec-plans/active/ to docs/exec-plans/completed/ and commit that move.
5. Create a pull request with:
   - Title: concise, under 70 characters
   - Body:
     ## Summary
     <2-4 bullet points>

     ## Milestones completed
     <from the plan>

     ## Key decisions
     <from the plan>

     ## Test plan
     <how to verify>

     Generated with Ralph Loop
6. Do not enable auto-merge and do not merge the PR.
7. Print the PR URL and report that the branch is ready for review.

Output <promise>COMPLETE</promise> when done.`, options.PlanPath, options.BaseBranch, options.BaseBranch)
}

File: apps/cli/crates/onequery-wtl/src/ralph/prompts.rs
use std::path::Path;

pub(crate) fn setup_developer_instructions() -> String {
    "You are the setup agent for Ralph Loop. Follow the prompt exactly and only output <promise>COMPLETE</promise> when the setup phase is fully done.".to_owned()
}

pub(crate) fn pr_developer_instructions() -> String {
    "You are the PR agent for Ralph Loop. Follow the prompt exactly and only output <promise>COMPLETE</promise> when the PR phase is fully done.".to_owned()
}

pub(crate) fn build_setup_prompt(
    user_prompt: &str,
    plan_path: &Path,
    worktree_path: &Path,
    worktree_id: &str,
    work_branch: &str,
    base_branch: &str,
) -> String {
    format!(
        concat!(
            "You are the setup agent for an automated coding loop. The worktree environment is already initialized.\n\n",
            "Task: {task}\n\n",
            "Prepared environment:\n",
            "- Worktree path: {worktree_path}\n",
            "- Worktree ID: {worktree_id}\n",
            "- Working branch: {work_branch}\n",
            "- Base branch: {base_branch}\n\n",
            "Do the following steps in order:\n\n",
            "1. Read `AGENTS.md`, `apps/cli/AGENTS.md`, and `apps/cli/docs/ARCHITECTURE.md`.\n",
            "2. Create an execution plan at `{plan_path}` using the checked-in plan structure.\n",
            "3. Break the work into 3-7 concrete milestones, each small enough for one coding-loop iteration.\n",
            "4. Mark every milestone as not started.\n",
            "5. Stage and commit the new plan with message: `plan: {plan_name}`.\n",
            "6. Print the absolute plan file path.\n\n",
            "Output <promise>COMPLETE</promise> when done."
        ),
        task = user_prompt,
        plan_path = plan_path.display(),
        plan_name = plan_path
            .file_stem()
            .and_then(|value| value.to_str())
            .unwrap_or("task"),
        worktree_path = worktree_path.display(),
        worktree_id = worktree_id,
        work_branch = work_branch,
        base_branch = base_branch
    )
}

pub(crate) fn build_pr_prompt(plan_path: &Path, base_branch: &str) -> String {
    format!(
        concat!(
            "You are a PR agent. Create a pull request for the completed work on this branch.\n\n",
            "Instructions:\n",
            "1. Read the completed plan at `{plan_path}`.\n",
            "2. Review commits with `git log {base_branch}..HEAD --oneline`.\n",
            "3. Review scope with `git diff {base_branch}...HEAD --stat`.\n",
            "4. Move the plan from `.worktree/plans/active/` to `.worktree/plans/completed/` and commit that move.\n",
            "5. Create a pull request with:\n",
            "   - Title: concise, under 70 characters\n",
            "   - Body sections: Summary, Milestones completed, Key decisions, Test plan, and `Generated with Ralph Loop`.\n",
            "6. Do not merge the PR yourself.\n",
            "7. Print the PR URL and report that the branch is ready for review.\n\n",
            "Output <promise>COMPLETE</promise> when done."
        ),
        plan_path = plan_path.display(),
        base_branch = base_branch
    )
}
